File: backend/app/services/geo.py
from math import atan2, cos, radians, sin, sqrt
from typing import Protocol


class TrackPoint(Protocol):
    latitude: float
    longitude: float


EARTH_RADIUS_METERS = 6_371_000


def haversine_distance_meters(
    first_latitude: float,
    first_longitude: float,
    second_latitude: float,
    second_longitude: float,
) -> float:
    lat_delta = radians(second_latitude - first_latitude)
    lon_delta = radians(second_longitude - first_longitude)
    first_lat = radians(first_latitude)
    second_lat = radians(second_latitude)

    value = sin(lat_delta / 2) ** 2 + cos(first_lat) * cos(second_lat) * sin(lon_delta / 2) ** 2
    return EARTH_RADIUS_METERS * 2 * atan2(sqrt(value), sqrt(1 - value))


def total_distance_meters(points: list[TrackPoint]) -> float:
    if len(points) < 2:
        return 0

    distance = 0.0
    for previous_point, current_point in zip(points, points[1:]):
        distance += haversine_distance_meters(
            previous_point.latitude,
            previous_point.longitude,
            current_point.latitude,
            current_point.longitude,
        )
    return round(distance, 2)

File: frontend/lib/geo.ts
import type { GpsPoint } from "@/lib/types";

const EARTH_RADIUS_METERS = 6_371_000;

export function haversineDistanceMeters(
  firstLatitude: number,
  firstLongitude: number,
  secondLatitude: number,
  secondLongitude: number,
) {
  const latDelta = toRadians(secondLatitude - firstLatitude);
  const lonDelta = toRadians(secondLongitude - firstLongitude);
  const firstLat = toRadians(firstLatitude);
  const secondLat = toRadians(secondLatitude);

  const value =
    Math.sin(latDelta / 2) ** 2 +
    Math.cos(firstLat) * Math.cos(secondLat) * Math.sin(lonDelta / 2) ** 2;

  return EARTH_RADIUS_METERS * 2 * Math.atan2(Math.sqrt(value), Math.sqrt(1 - value));
}

export function calculateDistanceMeters(points: Pick<GpsPoint, "latitude" | "longitude">[]) {
  if (points.length < 2) {
    return 0;
  }

  return Number(
    points
      .slice(1)
      .reduce((distance, point, index) => {
        const previousPoint = points[index];
        return (
          distance +
          haversineDistanceMeters(
            previousPoint.latitude,
            previousPoint.longitude,
            point.latitude,
            point.longitude,
          )
        );
      }, 0)
      .toFixed(2),
  );
}

export function formatDuration(totalSeconds: number) {
  const minutes = Math.floor(totalSeconds / 60);
  const seconds = totalSeconds % 60;
  return `${minutes}:${seconds.toString().padStart(2, "0")}`;
}

function toRadians(value: number) {
  return (value * Math.PI) / 180;
}
